remove QP's dependency on config

shyamjesal opened the pull request:
- Replace config file with environment variables
  - split env variables in 2 files: network and xdt 
  - Added struct tags in `utils/common.go` for default config params.
  - `utils/common.go` now parses environment variables using [envconfig](github.com/kelseyhightower/envconfig) instead of config file
  - Added environment variables in Makefile
  - Created `user-functions/gx/gx.yaml` and `user-functions/fx/fx.yaml` for knative deployment

- Removed redundent`ChunkSize` in `proto/crossXDT` `proto/downXDT`
- Manually append headers in http proxy as extracting metadata from `r.Context()` does not work
- Removed Payload un-marshalling at `sdk/golang/destination.go` as the metadata is send using http headers.
- Removed global config variable in `utils/common.go`
  - removed global config and handler variables in `sdk/golang/destination.go`
- Changed default routing mode to "CutThrough" 
- changed h2c ports to 50051 in fx.yaml and gx.yaml

shyamjesal committed:
removed payload deserialisation at the destination

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
remove QP's dependency on config.json

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

ustiugov reviewed (changes requested):
need to be re-done using env variables instead of headers. btw, you may want to keep you LoadConfig function but in the knative mode, it should parse the env variables. **Do not parse them in many places in the code.**

shyamjesal committed:
Load config variables from bash ENV

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
add environment file for docker compose

add knative yaml for fx and gx

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
fix config load in python SDK

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
fix fan-in-test data races

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
add env variables in benchmark test

remove config file

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

ustiugov commented on .github/workflows/xdt_tests.yml:
```diff
@@ -93,8 +93,26 @@ jobs:
         uses: actions/checkout@v2
       - name: Run benchmarks
+        env:
+          CHUNKSIZEINBYTES: 65536
```
it's not readable, please use underscores to separate words

ustiugov commented on user-functions/.env:
```diff
@@ -0,0 +1,18 @@
+CHUNKSIZEINBYTES=65536
```
what is this file? why don't makefiles use it?

shyamjesal replied:
this file is for loading env variables for docker compose run

ustiugov commented on user-functions/fx/fx.yaml:
```diff
@@ -0,0 +1,49 @@
+        - image: docker.io/vhiveease/fx:latest
+          imagePullPolicy: Always
+          env:
+            - name: CHUNKSIZEINBYTES
```
in the knative setup, hostnames and ports should not be configurable, they should have the default values.

please clear the redundant parameters

ustiugov commented on utils/common.go:
```diff
@@ -23,90 +23,50 @@
-	if err != nil {
-		log.Fatal(err)
-	}
+var LoadConfig = ReadConfig()
```
instead of a global variable, every component should call ReadConfig(). Globals are bad style

ustiugov reviewed (changes requested):
- [x] use underscores for env variables for readability
- [x] why routing is still a runtime argument? it should be via env vars
- [x] in the knative setup, hostnames and ports should not be configurable, they should have the default values. please remove redundant parameters
- [x] remove the global LoadConfig variable

shyamjesal committed:
improved readability of env variables

moved benchmark test to makefile
removed hostname and port config variables from knative yamls

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
removed config global variable

removed global config and handler variables in destination.go

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
fix docker-compose

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

ustiugov commented on queue-proxy/sQP/sQP.go:
```diff
@@ -61,11 +61,13 @@ func (s upXDTServer) SendData(srv upXDT.StreamData_SendDataServer) error {
 		if err == io.EOF {
 			log.Infof("sQP: %d chunks received", chunkCount)
-			if config.Routing == utils.STORE_FORWARD {
+			if headers["routing"][0] == utils.STORE_FORWARD {
```
this is what I don't get. why is routing in the header? it should be upon startup with an env variable, right?

ustiugov commented on sdk/golang/destination.go:
```diff
@@ -39,50 +40,47 @@ import (
-	key := xdtPayload.Key
+	if ok && headers["is_xdt"][0] == "true" {
+		key := headers["key"][0]
+		log.Infof("DST: using %s routing", headers["routing"][0])
```
shouldn't be in headers

ustiugov commented on sdk/python/source.py:
```diff
@@ -53,32 +53,32 @@ def splitPayload(xdtPayload):
         ('is_xdt', 'true'),
         ('key', key),
         ('sqp_addr', sQPAddr),
+        ('routing', config['Routing']),
```
shouldn't be here

ustiugov commented on sdk/python/test.py:
```diff
@@ -39,23 +39,27 @@ def test_splitPayload(self):
+            ('is_xdt', 'true'),
+            ('key', 'secret'),
+            ('sqp_addr', config['SQPServerHostname']+config['SQPServerPort']),
+            ('routing', config['Routing']),
```
shouldn't be in a header

ustiugov commented on tests/integration_test.go:
```diff
@@ -96,27 +99,33 @@ func knativeQP(config utils.Config) {
+					"is_xdt":   r.Header.Get("is_xdt"),
+					"key":      r.Header.Get("key"),
+					"sqp_addr": r.Header.Get("sqp_addr"),
+					"routing":  r.Header.Get("routing"),
```
shouldn't be in a header

ustiugov commented on user-functions/dQP/dQP.go:
```diff
@@ -82,18 +83,26 @@ func main() {
+					"is_xdt":   r.Header.Get("is_xdt"),
+					"key":      r.Header.Get("key"),
+					"sqp_addr": r.Header.Get("sqp_addr"),
+					"routing":  r.Header.Get("routing"),
```
shouldn't be in a header

ustiugov commented on Makefile:
```diff
@@ -29,6 +29,25 @@ ROUTING_MODES:=Store&Forward CutThrough
 GO_TEST_FLAGS:=-race -v -cover
 SDK_GO_FILES:=./source.go ./destination.go
 
+.EXPORT_ALL_VARIABLES:
```
is it possible to import it from the .env file that you use for docker compose? 2 sets of vars is hard to maintain

ustiugov commented on user-functions/fx/fx.yaml:
```diff
@@ -0,0 +1,33 @@
+              value: "10"
+          ports:
+            - name: h2c
+              containerPort: 50010
```
please use the same 50051 port in knative in all functions

ustiugov commented on user-functions/gx/gx.yaml:
```diff
@@ -0,0 +1,33 @@
+              value: "10"
+          ports:
+            - name: h2c
+              containerPort: 50007
```
50051 please

shyamjesal committed:
split .env file into network and xdt

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
fix misaligned env variables in fx and gx yamls

changed h2c ports to 50051 in fx.yaml and gx.yaml

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>

shyamjesal committed:
change default routing to CutThrough

Signed-off-by: Shyam Jesal <s.jesalpura@gmail.com>
